### PR TITLE
AutoFlow

### LINKED ISSUE #1570: Automated workflows / PDF report generation
Add a new FlowKit component that sits outside the API, and can be used to automate running parametrised workflows involving FlowAPI queries for each date of CDR data added to FlowDB.

This component should include functionality to generate  PDF reports as outputs of an automated workflow.

### PR DESCRIPTION
**Pull request by jch1g10**
Closes #1570

### I have:

- [x] Formatted any Python files with [black](https://github.com/ambv/black)
- [x] Brought the branch up to date with master
- [x] Added any relevant Github labels
- [x] Added tests for any new additions
- [x] Added or updated any relevant documentation
- [x] Added an Architectural Decision Record (ADR), if appropriate
- [x] Added an [MPLv2 License Header](https://www.mozilla.org/en-US/MPL/headers/) if appropriate
- [x] Updated the [Changelog](https://github.com/Flowminder/FlowKit/blob/master/CHANGELOG.md) 

### Description

Adds a new component, AutoFlow, to FlowKit.

AutoFlow is a tool that automates the event-driven execution of workflows consisting of Jupyter notebooks that interact with FlowKit via FlowAPI. Workflows can consist of multiple inter-dependent notebooks, and can run automatically for each new date of CDR data available in FlowDB. After execution, notebooks can optionally be converted to PDF reports.

AutoFlow uses:
- [Prefect](https://github.com/prefecthq/prefect) to define and run workflows,
- [Papermill](https://github.com/nteract/papermill) to parametrise and execute Jupyter notebooks,
- [Scrapbook](https://github.com/nteract/scrapbook) to enable data-sharing between notebooks,
- [nbconvert](https://github.com/jupyter/nbconvert) and [asciidoctor-pdf](https://github.com/asciidoctor/asciidoctor-pdf) to convert notebooks to PDF, via asciidoc.

There are two categories of workflow used within AutoFlow:
- Notebook-based workflow: A Prefect flow that executes one or more Jupyter notebooks using Papermill, and optionally converts the executed notebooks to PDF. These are created using the function `autoflow.workflows.make_notebooks_workflow`.
- Sensor: A Prefect flow that runs on a schedule, checks for new events, and runs a set of notebook-based workflows for each event for which they haven't previously run successfully. One sensor is currently implemented - `autoflow.sensor.available_dates_sensor`, which uses the FlowAPI `get_available_dates` route to get a list of available dates of CDR data, and runs notebook-based workflows for each new date found.

Available dates sensor flow:
available_dates_sensor

AutoFlow is designed to run in a Docker container with two bind mounts:
- Inputs directory (mounted read-only), which contains the notebooks to be executed along with a YAML file `workflows.yml` that defines the workflows to execute the notebooks and configures the available dates sensor to run these workflows.
- Outputs directory, which will contain two sub-directories containing the executed notebooks and PDF reports, respectively.

### Example
`autoflow/examples/` contains an example AutoFlow workflow, which runs two Jupyter notebooks and produces a PDF report showing in/outflows above normal for each day of CDR data. `autoflow/examples/inputs/workflows.yml` configures the available dates sensor to check for new dates every day at midnight, and run the flows-above-normal workflow with two different sets of parameters.

The AutoFlow example can be run from the FlowKit root by first creating a FlowAPI token, and then running
```bash
set -a && . development_environment && set +a
export FLOWAPI_TOKEN=<token>
make up DOCKER_SERVICES="flowdb_testdata flowmachine flowmachine_query_locker flowapi autoflow"
```
(To run the sensor immediately without a schedule, rather than waiting until midnight, change the line `schedule: "0 0 * * *"` to `schedule: null` in `autoflow/examples/inputs/workflows.yml`).

Outputs will be in `autoflow/examples/outputs/`.

### CODE REVIEW

File: autoflow/autoflow/date_stencil.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    raw_stencil : sequence of date, int and/or pairs of date/int
+        List of elements defining date intervals.
+        Each element can be:
+            - a date object corresponding to an absolute date,
+            - an int corresponding to an offset (in days) relative to a reference date,
+            - a length-2 sequence [start, end] of dates or offsets,
+              corresponding to a date interval (inclusive of both limits).
+    """
+
+    def __init__(
```
Comment (greenape) [inline autoflow/autoflow/date_stencil.py]: I think a better way of passing these params and creating the object can be found. Not vital right now, but this is a little awkward.

File: autoflow/autoflow/parser.py [hunk 2]
```diff
@@ -0,0 +1,450 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+"""
```
Comment (greenape) [inline autoflow/autoflow/parser.py]: Break this up into file-per-class I think.

File: autoflow/autoflow/parser.py [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+    output = fields.Nested(NotebookOutputSchema, required=False)
+
+    @validates("filename")
+    def validate_filename(self, value):
+        """
+        Raise a ValidationError if the specified file does not exist or is not a Jupyter notebook,
+        or if the inputs directory is not specified in the context.
+        """
+        if Path(value).suffix != ".ipynb":
+            # TODO: '.ipynb' does not a notebook make
```
Comment (greenape) [inline autoflow/autoflow/parser.py]: Not a big deal to my mind - you could validate that it is a true notebook by parsing the json, but that's a _lot_ of effort that doesn't really gain you anything.

File: autoflow/autoflow/parser.py [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+    
+    Returns
+    -------
+    workflows : list of Flow
+        List of prefect workflows
+    run_parameters : dict
+        mapping from workflow names to a list of dicts of parameters for which the workflow should be run
+    """
+    with open(Path(inputs_dir) / filename, "r") as f:
+        workflows_yaml = yaml.safe_load(f)
```
Comment (greenape) [inline autoflow/autoflow/parser.py]: Kind of a drag that two config formats are used for the one task?

File: autoflow/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    && rm -rf \
+    /var/lib/apt/lists/* \
+    /tmp/* \
+    /var/tmp/* \
+    /usr/share/man \
+    /usr/share/doc \
+    /usr/share/doc-base
+
+WORKDIR /${SOURCE_TREE}/autoflow
+RUN pip install -U pip && pip install .[postgres,examples]
```
[the remark this answers is not in the document]
Comment (jch1g10) [reply autoflow/Dockerfile]: Yes, probably better to use the Pipfile, I guess.

File: autoflow/autoflow/date_stencil.py (hunk 1 above)
Comment (jch1g10) [reply autoflow/autoflow/date_stencil.py]: Agreed. Originally I was using the 'raw' format throughout, before I added the `DateStencil` class, and I think a little more refactoring is still needed here.

File: autoflow/autoflow/parser.py (hunk 4 above)
Comment (jch1g10) [reply autoflow/autoflow/parser.py]: This and the Prefect `config.toml`, you mean? I think the Prefect config setup needs re-working a bit anway - `config.toml` doesn't allow for default settings if an environment variable is missing, and can't use Docker secrets, so I'd like to find a better solution there.

**Pull request merged**